experiment/simulation/js/gate.js

```diff
@@ -0,0 +1,632 @@
+        x = (x / scale) * width;
+        y = (y / yScale) * height;
+
+        const el = document.getElementById(this.id);
+        el.style.left = x + "px";
+        el.style.top = y + "px";
+
+        if (this.type != "Input" && this.type != "Output") {
+
+            el.addEventListener(
+                "contextmenu",
+                function (ev) {
+                    ev.preventDefault();
+                    const origin = {
+                        left: ev.pageX - document.getScroll()[0],
+                        top: ev.pageY - document.getScroll()[1],
+                    };
+                    setPosition(origin);
+                    window.selectedComponent = this.id;
+                    window.componentType = "gate";
+                    return false;
+                },
+                false
+            );
+        }
+        gates[this.id] = this;
+        registerGate(this.id, this);
+
+        this.updatePosition(this.id);
+    }
+
+    addInputPoints(input) {
+        this.inputPoints.push(input);
+    }
+
+    addOutputPoints(output) {
+        this.outputPoints.push(output);
+    }
+
+
+    generateOutput() {
+        switch (this.type) {
+            case "AND":
+                this.output = computeAnd(getOutput(this.inputs[0]), getOutput(this.inputs[1]));
+                break;
+            case "OR":
+                this.output = computeOr(getOutput(this.inputs[0]), getOutput(this.inputs[1]));
+                break;
+            case "NOT":
+                this.output = !getOutput(this.inputs[0]);
+                break;
+            case "NAND":
+                this.output = computeNand(getOutput(this.inputs[0]), getOutput(this.inputs[1]));
+                break;
+            case "NOR":
+                this.output = computeNor(getOutput(this.inputs[0]), getOutput(this.inputs[1]));
+                break;
+            case "XOR":
+                this.output = computeXor(getOutput(this.inputs[0]), getOutput(this.inputs[1]));
+                break;
+            case "XNOR":
+                this.output = computeXnor(getOutput(this.inputs[0]), getOutput(this.inputs[1]));
+                break;
+            case "ThreeIPNAND":
+                this.output = !(getOutput(this.inputs[0]) && getOutput(this.inputs[1]) && getOutput(this.inputs[2]));
+                break;
+            case "Output":
+                this.output = getOutput(this.inputs[0]);
+                break;
+        }
+    }
+
+    setOutput(val) {
+        this.output = val;
+    }
+    setConnected(val) {
+        this.isConnected = val;
+    }
+}
+
+function getOutput(input) {
+    let gate = input[0];
+    let pos = input[1];
+
+    if (pos === EMPTY) {
+        return gate.output;
+    }
+    else if (pos === "Q") {
+        return gate.q;
+    }
+    else if (pos === "Q'") {
+        return gate.qbar;
+    }
+}
+
+
+function addGate(event) {
+    let type = event.target.innerHTML;
+    if (type === "3-NAND")

```

Comment:
Please use a block {} for the if statement even if there's a single statement inside it

Reply:
done